From ff9a9ac11552de274f4064a2d326fb746c38a23c Mon Sep 17 00:00:00 2001
From: Mel Gorman <mgorman@suse.de>
Date: Thu, 27 Aug 2015 10:11:23 +0100
Subject: [PATCH] mm: filemap: Avoid unnecessary barriers and waitqueue lookups
 -fix

References: VM/FS Performance (bnc#941951)
Patch-mainline: No, upstream resisting until every other problem solved first

This patch addresses a problem wherby a call to __ClearPageWaiters was
missed in a rare case where a page was not successfully invalidated.

Signed-off-by: Mel Gorman <mgorman@suse.de>
---
 mm/vmscan.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/mm/vmscan.c b/mm/vmscan.c
index fb5e4b0ff99b..640d8d6540c0 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -1085,9 +1085,10 @@ static unsigned long shrink_page_list(struct list_head *page_list,
 		 */
 		__clear_page_locked(page);
 
+free_it:
 		/* See release_pages on why this clear may be necessary */
 		__ClearPageWaiters(page);
-free_it:
+
 		nr_reclaimed++;
 
 		/*
